drivers/ipm/ipm_stm32_hsem.c

```diff
@@ -0,0 +1,197 @@
+{
+	struct stm32_hsem_mailbox_data *data = dev->data;
+
+	ARG_UNUSED(wait);
+	ARG_UNUSED(buff);
+
+	if (size) {
+		LOG_WRN("stm32 HSEM not support data transfer");
+		return -EMSGSIZE;
+	}
+
+	if (id) {
+		LOG_WRN("stm32 HSEM only support a single instance of mailbox");
+		return -EINVAL;
+	}
+
+	/* Lock the semaphore tx_semid */
+	z_stm32_hsem_lock(data->tx_semid, HSEM_LOCK_DEFAULT_RETRY);
+
+	/**
+	 * Release the semaphore tx_semid.
+	 * This will trigger a HSEMx interrupt on another CPU.
+	 */
+	z_stm32_hsem_unlock(data->tx_semid);
+
+	return 0;
+}
+
+void stm32_hsem_mailbox_ipm_register_callback(const struct device *dev,
+					ipm_callback_t cb,
+					void *user_data)
+{
+	struct stm32_hsem_mailbox_data *data = dev->data;
+
+	data->callback = cb;
+	data->user_data = user_data;
+}
+
+int stm32_hsem_mailbox_ipm_max_data_size_get(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+    /* stm32 HSEM not support data transfer */
+	return 0;
+}
+
+uint32_t stm32_hsem_mailbox_ipm_max_id_val_get(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	/* stm32 HSEM only support a single instance of mailbox */
+	return 0;
+}
+
+int stm32_hsem_mailbox_ipm_set_enabled(const struct device *dev, int enable)
+{
+	struct stm32_hsem_mailbox_data *data = dev->data;
+	uint32_t mask_semid = (1U << data->rx_semid);
+
+	if (enable) {
+		/* Clear semaphore rx_semid interrupt status and masked status */
+		LL_HSEM_ClearFlag_CICR(HSEM, mask_semid);
+		/* Enable semaphore rx_semid on HESMx interrupt */
+		LL_HSEM_EnableIT_CIER(HSEM, mask_semid);
+	} else {
+		/* Disable semaphore rx_semid on HSEMx interrupt */
+		LL_HSEM_DisableIT_CIER(HSEM, mask_semid);
+	}
+
+	return 0;
+}
+
+static int stm32_hsem_mailbox_init(const struct device *dev)
+{
+	struct stm32_hsem_mailbox_data *data = dev->data;
+	const struct stm32_hsem_mailbox_config *cfg = dev->config;
+	const struct device *clk = DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE);
+
+	/* Enable clock */
+	if (clock_control_on(clk, (clock_control_subsys_t *)&cfg->pclken) != 0) {
+		return -EIO;
+	}
+
+	/* Config transfer semaphore */
+	if (CONFIG_IPM_STM32_HSEM_CPU == HSEM_CPU1) {
+		data->tx_semid = CFG_HW_IPM_CPU2_SEMID;
+		data->rx_semid = CFG_HW_IPM_CPU1_SEMID;
+	} else {
+		data->tx_semid = CFG_HW_IPM_CPU1_SEMID;
+		data->rx_semid = CFG_HW_IPM_CPU2_SEMID;
+	}
+
+	cfg->irq_config_func(dev);
+
+	return 0;
+}
+
+static const struct ipm_driver_api stm32_hsem_mailbox_ipm_dirver_api = {
+	.send = stm32_hsem_mailbox_ipm_send,
+	.register_callback = stm32_hsem_mailbox_ipm_register_callback,
+	.max_data_size_get = stm32_hsem_mailbox_ipm_max_data_size_get,
+	.max_id_val_get = stm32_hsem_mailbox_ipm_max_id_val_get,
+	.set_enabled = stm32_hsem_mailbox_ipm_set_enabled,
+};
+
+static const struct stm32_hsem_mailbox_config stm32_hsem_mailbox_0_config = {
+	.irq_config_func = stm32_hsem_mailbox_irq_config_func,
+	.pclken = {
+		.bus = DT_INST_CLOCKS_CELL(0, bus),
+		.enr = DT_INST_CLOCKS_CELL(0, bits)
+	},
+};
+
+DEVICE_DT_INST_DEFINE(0,
```

Comment:
Is there a good reason why this does not respect `status = "disabled";` by using DT_INST_FOREACH_STATUS_OKAY?

Reply:
STM32 HSEM has its own LL_HSEM(low-level HSEM) API provided by the hal_stm32 module. LL_HSEM API worked very well for the hardware semaphore mechanism.
The STM32 HSEM IPM driver I add here, Only pick up two semaphore ID from stm32_hsem.h to simulate a virtual mailbox device. 
So there will have only one instance.

Reply:
Having a single instance does not excuse ignoring status 🙂. I think the default stance should be to respect it.

Reply:
Sorry, misunderstanding the question, will add the 'status' check.

Reply:
I have added the DT_INST_FOREACH_STATUS_OKAY.